libgnucash/engine/mocks/gmock-Split.cpp

```diff
@@ -30,113 +30,143 @@ gnc_mock_split_class_init (MockSplitClass *klass)
 Split *
 xaccMallocSplit (QofBook *book)
 {
-    g_return_val_if_fail(QOF_IS_MOCK_BOOK(book), NULL);
-    return ((QofMockBook*)book)->mallocSplit();
+    EXPECT_TRUE(QOF_IS_MOCK_BOOK(book));
+    if (QOF_IS_MOCK_BOOK(book))
+        return ((QofMockBook*)book)->malloc_split();
+    else
+        return nullptr;
 }
 
 QofBook *
 xaccSplitGetBook (const Split *split)
 {
-    g_return_val_if_fail(GNC_IS_MOCK_SPLIT(split), NULL);
-    return ((MockSplit*)split)->getBook();
+    EXPECT_TRUE(GNC_IS_MOCK_SPLIT(split));
+    if (GNC_IS_MOCK_SPLIT(split))
+        return ((MockSplit*)split)->get_book();
+    else
+        return nullptr;
```

Comment:
It's [conventional for `GObjects`](https://developer.gnome.org/gobject/stable/gtype-conventions.html) to have a macro to do this, in this case named `GNC_MOCK_SPLIT`. Conventionally the macro would be implemented as
```C
#define GNC_MOCK_SPLIT (split) \
   g_return_val_if_fail (GNC_IS_MOCK_SPLIT (split), NULL)
```
so that it logs an error as well, but I think that would be undesirable in test code.

Reply:
Never mind, that's not what you're doing.

Reply:
I already replaced `g_return_if_fail()` and `g_return_val_if_fail()` by `ASSERT_TRUE` and `EXPECT_TRUE` in commit bb0088c, since you mentioned this in PR #650. That means `g_return_val_if_fail()` has to be replaced by `if GNC_IS_MOCK_SPLIT (split) { return NULL }` in this macro?

What about cases with return values other than `NULL` or `nullptr` in case of incorrect type?

Shouldn't this macro also return a type-casted pointer?

Reply:
I didn't show the whole macro. The whole thing is
```C
#define GNC_MOCK_SPLIT (split)                                        \
{                                                                                              \
    g_return_val_if_fail (GNC_IS_MOCK_SPLIT (split), NULL); \
    return (MockSplit*)split;                                                              \
}
```
The point being to create a checked typecast. That's necessary because C is a bit loose on type enforcement for pointer types and doesn't really understand types beyond its POD ones. So GObject has a run-time type system that requires runtime checks and carefully nested structs to provide a semblance of polymoprhism. C++ has all of that built into the language, understands types very deeply, and enforces good type safety at compile time. That's been tightened up in the last 10 years. "NULL" is a C hack: In the C headers somewhere you'll find `#define NULL (void*)0`, while in C++11 and later `nullptr` is a language keyword that is typeless, the intent being to (and soon, we just got a bug that gcc11 enforces this) make it illegal to return `void*` when the function signature is `foo*`.

Reply:
I'm a bit confused. In PR #650 you commented code line
```C
g_return_if_fail(GNC_IS_MOCK_ACCOUNT(account));
```
with

> ISTM you should use EXPECT_TRUE() or maybe ASSERT_TRUE instead of g_return_if_fail. The latter will just write a message to std::cerr instead of causing the test to fail.

Therefore I replaced all `g_return_if_fail()` with `ASSERT_TRUE` in commit bb0088c, which is straightforward, if the current function has no return value ([`ASSERT_TRUE`](https://github.com/google/googletest/blob/release-1.8.1/googletest/docs/primer.md#basic-assertions) is a fatal assertion, i.e. it immediately aborts the current function, which is similar to the behaviour of `g_return_if_fail()`.
But did this comment also mean, that all `g_return_val_if_fail()` should be replaced by `EXPECT_TRUE`? This is, what I did as well in commit bb0088c. This contradicts to your proposal to use `g_return_val_if_fail()` in macros like `GNC_MOCK_SPLIT`. Or should `g_return_val_if_fail()` only be complemented with `EXPECT_TRUE`?

Would it make sense to define macro `GNC_MOCK_SPLIT` as
```C
#define GNC_MOCK_SPLIT (split)                              \
{                                                           \
    EXPECT_TRUE(GNC_IS_MOCK_SPLIT(split));                  \
    g_return_val_if_fail (GNC_IS_MOCK_SPLIT(split), NULL);  \
    return (MockSplit*)split;                               \
}
```
or
```C
#define GNC_MOCK_SPLIT (split)                              \
{                                                           \
    EXPECT_TRUE(GNC_IS_MOCK_SPLIT(split));                  \
    if (GNC_IS_MOCK_SPLIT(split))                           \
        return (MockSplit*)split;                           \
    else                                                    \
        return NULL;                                        \
}
```
?

Reply:
Yes, `EXPECT_TRUE` is better that `g_return_if_fail`. `EXPECT_TRUE` will fail the test, `g_return_if_fail` won't. You might do instead
```
static inline MockSplit*
mock_split(Split* split)
{
    if (GNC_IS_MOCK_SPLIT(split))
           return static_cast<MockSplit*>(split);
    ADD_FAILURE() << "Not a MockSplit!";
    return nullptr;
}
```
So you need do the test only once. This is C++, better to not use the preprocessor. BTW you'll need a second one for `const Split*` -> `const MockSplit*`.

I decided to build a simple one-function test of Account using MockSplit and QofMockBook so I could play around with this a bit. You can see the result at https://github.com/jralls/gnucash/tree/gmock-tryout. I had to spend a fair amount of time building out the mocked functions and I opted for not doing checks in them. But even though I know GnuCash is spaghetti like I was surprised how long it took to collect all of the source files needed to compile Account.cpp (56, not including the mocks).

Some important lessons:
* The C replacement function definitions need to be wrapped in `extern "C" {...}` unless the real implementation file is C++; otherwise they get C++ linkage and the linker won't recognize them.
* Functions that take a `const` first argument need to be mocked with `MOCK_CONST_METHOD`.
* `foo_get_type` needs to be faked to `return mock_foo_get_type()` so that the runtime GType checking in the code under test succeeds.
* Runtime checks of other classes in mock code should check for the base-class type, not for its mock-class if it has one. Mocks shouldn't assume that they'll be used only with the other mocks.

Reply:
Using `ADD_FAILURE()` to avoid testing twice is a good idea. I'll implement it this way.

But do you also mean to implement the type conversion as  a function, `mock_split()` in this case? I see this in your code on branch `gmock-tryout`. Or would it still be sufficient to write a macro `GNC_MOCK_SPLIT` with the content of your example `mock_split()` according to the GObject conventions?

If a function is intended, it should better be named `gnc_mock_split()`, I guess.

>  BTW you'll need a second one for const Split* -> const MockSplit*.

How is this done in other GObject code, if type conversion is implemented as a macro?

>  The C replacement function definitions need to be wrapped in `extern "C" {...}` unless the real implementation file is C++; otherwise they get C++ linkage and the linker won't recognize them.

The *definitions*, not the *declarations* only? So it isn't sufficient to wrap `#include <Split.h>` and `#include <SplitP.h>` in *gmock-Split.h*? Why is the test application `test-import-backend` already linked successfully then? I don't have precise knowledge about C++ linkage.

Reply:
Preprocessor macros are *very* strongly discouraged in C++, so yes, an `inline static` function is better. GObject is in C and abuses macros and other C features rather heavily in its lame efforts to replicate C++. It's something to use when one must but not something to emulate.

Why would you name the class `MockSplit` and want to call the casting function `gnc_mock_split`? 

> I don't have precise knowledge about C++ linkage.

You need to learn about that, it's important when integrating C and C++. 

The short explanation is that C++ "mangles" function names with information about their namespaces and parameters, which is necessary to make namespaces and overloading work. C doesn't, so in order to be able to use a C++ function from C you have to tell the compiler not to mangle it by declaring it `extern "C"`. Marking the decls only works if the marked decls are the first ones for the function that the compiler sees because the compilation guards prevent the compiler from looking inside subsequent ones. The compiler will also raise an error if it encounters multiple decls with differing signatures and `extern "C"` is part of the signature. 

> Why is the test application test-import-backend already linked successfully then?

Because it's all C++.

Reply:
>  Why would you name the class `MockSplit` and want to call the casting function `gnc_mock_split`?

Because of the GObject type name `gnc_mock_split`, which is used for the GObject related functions `gnc_mock_split_init()` and `gnc_mock_split_class_init()` already.

I have another question related to your code on branch `gmock-tryout`. In contrast to your definition of function `mock_split()` here (see above) you don't do any GObject runtime type check on branch `gmock-tryout` before type conversion. This is only left out for your private use, right?

>  Marking the decls only works if the marked decls are the first ones for the function that the compiler sees because the compilation guards prevent the compiler from looking inside subsequent ones.

But doesn't this mean, that it's sufficient to mark the declarations only, since the header file containing the declarations is included in the cpp file before the definitions? And do I understand https://en.cppreference.com/w/cpp/language/language_linkage correctly, that using notation `extern "C" { ... }` marks the declaration and the definition, even if only the declaration is contained inside the braces?

Reply:
> Because of the GObject type name gnc_mock_split

Then the class should be GncMockSplit.

If we were using gobject-introspection (be glad that we don't!) the class would have to be GncMocksplit and the functions gnc_mocksplit_ for the scanner to be able to correctly separate the namespace, class, and function names.

> I have another question related to your code on branch gmock-tryout...

Not entirely. I was also experimenting with whether the compiler would protect the mock code from getting a non-Split*. It does as long as everything is C++ so it works in this case and would as well in the tests you wrote. It's not safe generally so the checks should go in.

> But doesn't this mean, that it's sufficient to mark the declarations only, since the header file containing the declarations is included in the cpp file before the definitions? And do I understand https://en.cppreference.com/w/cpp/language/language_linkage correctly, that using notation extern "C" { ... } marks the declaration and the definition, even if only the declaration is contained inside the braces?

You do understand the cppreference.com item correctly. What you're missing is that just because `#include <Split.h>` is in this file before the definitions doesn't mean that that particular include is the one that actually includes the declarations. If `Split.h` is already indirectly included then the compile guard has already been invoked and the later include doesn't do anything.

I think that's actually not a problem for MockSplit. It *was* a problem for QofBook because Qof has a rather convoluted inclusion arrangement thanks to its time as a separate library.

Reply:
>  Then the class should be GncMockSplit.

And why is the GObject class `Split` defined as follows then?
```
G_DEFINE_TYPE(Split, gnc_split, QOF_TYPE_INSTANCE)
```
It's the same naming scheme. The camel case type name is `Split` and the lowercase type name is `gnc_split`.

>  If Split.h is already indirectly included then the compile guard has already been invoked and the later include doesn't do anything.

That's it, I got it.

Reply:
Most of the split functions are named xaccSplitFoo because Split is one of the original X-Accounting classes from before Gnome. The port to GObject came almost 10 years later and when @derekatkins [converted it to GObject](https://github.com/Gnucash/gnucash/commit/cb8f58790c3a59f3236d3f66adfc99ef97241695) he didn't change the name. BTW he often says he regrets having done that instead of making them C++ classes.

Reply:
Interesting. This means that `Split` should have been actually named `GncSplit`? GObject recommends to use prefixes as namespaces, which is `Gnc` or `gnc_` here.

And if I understand you correctly, the prefix should be omitted for the mock classes, since it's C++. But this means, the lower case type name of `MockSplit` should be changed to `mock_split` instead of `gnc_mock_split`, right?

Reply:
> Sorry, I ask again regarding placement of extern "C". Wouldn't it be the best to wrap the declarations directly in the C headers together with the #ifdef __cplusplus guard? This would avoid the described problem, that someone could include Split.h for instance without extern "C" wrapper by accident.

Yes, we'll need to do that sooner or later, though accidents of that sort aren't to worrying as they're caught at link time.

Reply:
> Interesting. This means that Split should have been actually named GncSplit?

Well, renamed that way.

> And if I understand you correctly, the prefix should be omitted for the mock classes, since it's C++.

It would be more about the mock classes being for testing and therefore not exported out to where namespaces matter. Given that C++ has real namespaces and that we ultimately want to remove GObject from everywhere that doesn't really need it, we'd end up with something like `Gnc::Engine::Split` and `Gnc::Engine::Mock::Split`. Once Split is a C++ class we'll also be able to dispense with all of this GObject runtime crap. The compiler will make sure that we get a Split(-subclass) in a function that takes one.

>  But this means, the lower case type name of MockSplit should be changed to mock_split instead of gnc_mock_split, right?

Depends. If you're following GObject rules the type name would be `GncMocksplit` and functions would be named `gnc_mocksplit_foo`. There are two reasons to follow those rules: One is that you're using gobject-introspection, which needs things formatted that way in order to create the object->function names for languages that work that way. We don't, so that's out. The other would be that you're using GObject's built in lifetime management: The macros that manage that stuff also need names that follow those rules, but not quite as strictly. The original X-accounting classes don't use those macros and so don't follow the rules, the Qof and business classes do use the macros and so they have names like QofInstance and GncInvoice.

There's no point in any `gnc_mock_foo_` function of any kind, it's never called. The mocks have to provide all of the `gnc_foo_` functions that are actually used in the code being tested so that it will link. Those can be implemented in C++, just as you already did with the mocked functions. The only reason to provide for the lifecycle functions (`gnc_foo_init`, `gnc_foo_class_init`, `gnc_foo_dispose`, and `gnc_foo_finalize`) is if the function under test creates one. If you're going to create the object in `SetUp` and destroy it in `TearDown` then it can be a pure C++ object.
Things get a bit more interesting when one wants to test QofInstance behavior (e.g. `qof_instance_get(QOF_INSTANCE(split))` )  through one of the classes, but let's leave that for later work.

Reply:
I did part of the suggested changes now.

Next I wanted to implement the following suggestion:
> Runtime checks of other classes in mock code should check for the base-class type, not for its mock-class if it has one. Mocks shouldn't assume that they'll be used only with the other mocks.

This led me to the problem, that checking for the base-class type, for instance `GNC_TYPE_ACCOUNT`, requires the base-class to be always linked to the test application. In this example the corresponding function `gnc_account_get_type()` is defined in *Account.cpp*, which has be always linked to the application in this case. But don't we run again into the same problem, we discussed in PR #650 already? We wanted to completely **replace** `libgncmod-engine.so` by mockups to not get in conflict with doubled defined symbols. Linking *Account.cpp* contradicts to that decision. By the way, that decision is the reason why `GNC_TYPE_MOCKACCOUNT` is not derived from `GNC_TYPE_ACCOUNT` but replaces it instead.

On the other hand you furthermore suggested

> `foo_get_type` needs to be faked to return `mock_foo_get_type()` so that the runtime GType checking in the code under test succeeds.

Isn't that the same as using `GNC_IS_MOCKACCOUNT()` instead of `GNC_IS_ACCOUNT()`?

Reply:
> This led me to the problem, that checking for the base-class type, for instance GNC_TYPE_ACCOUNT, requires the base-class to be always linked to the test application.

You missed from [gmopck-tryout](https://github.com/jralls/gnucash/commit/088a53ea6d37ed884408822b930384dbc38b392b)
```C++


GType gnc_split_get_type(void)
{
    return gnc_mock_split_get_type();
}
```

Reply:
> Isn't that the same as using GNC_IS_MOCKACCOUNT() instead of GNC_IS_ACCOUNT()?

Nope. With `gnc_split_get_type` returning `gnc_mocksplit_get_type` all of the type macros become interchangeable. That's important if you're going to pass a MockSplit to a C function taking a Split.

Reply:
Fakes for `_get_type()` functions added and runtime type checks changed.

The last issue, adding `extern "C"` for C linkage, still confuses me. I'll update that later.

Reply:
Ok, I now also added missing C linkage statements to the best of my knowledge and belief.

But there are differences to your `gmock-tryout` branch:
- I didn't wrap *gmock-Split.cpp* and only wrapped `#include <SplitP.h>` in *gmock-Split.h*, since *Split.h* already contains `extern "C"`
- I didn't wrap *gmock-qofbook.cpp*, since *qofbook.h* and *qofbook-p.h* already contain `extern "C"`